CB-251: Reworked WS.java

jmittert opened the pull request:
* Added support for sending JSON requests
* Updated to use the latest Apache httpclient functions
* The send body is no longer always null
* Now handles headers properly

Test run against grimoire with 14 failures, production has 24.
http://ci/view/Voodoo/job/Voodoo_CB_Custom_Tests/88/
http://ci/view/Voodoo/job/Voodoo_CB_Custom_Tests/89/

http://ci/job/7700beta6_Voodoo_Self_Tests/28/

One failure was due to a chromedriver failure, the other due to an unrelated issue.

Eriten commented on src/main/java/com/sugarcrm/candybean/webservices/WS.java:
```diff
@@ -21,216 +21,239 @@
+		if (body == null) {
+			return request(op, uri, headers, payload);
+		} else {
+			return request(op,uri, headers, body, ContentType.DEFAULT_TEXT);
```
Inconsistent format! space after comma

Eriten commented on src/main/java/com/sugarcrm/candybean/webservices/WS.java:
```diff
@@ -21,216 +21,239 @@
+	 * @throws Exception If HTTP request failed
+	 */
+	public static Map<String, Object> request(OP op, String uri, Map<String, String> headers, Map<String,String> body) throws Exception {
+		return request(op,uri,headers,body,ContentType.APPLICATION_JSON);
```
Inconsistent format! space after comma

jmittert commented:
Cleaned up several whitespace issues.

Eriten commented on src/main/java/com/sugarcrm/candybean/webservices/WS.java:
```diff
@@ -5,232 +5,251 @@
-				request = put;
-	            mapParse = requestIt(request, headers);
-				break;
+				return handleRequest(put, headers);
```
The PUT and POST pretty much have the same code inside the case. We should refactor this. Put the code out and put it in a private method and you can represent `HttpPut` and `HttpPost` as `HttpEntityEnclosingRequestBase`.

Eriten commented on src/main/java/com/sugarcrm/candybean/webservices/WS.java:
```diff
@@ -5,232 +5,251 @@
+	 * @return Key Value pairs of the response
+	 * @throws Exception If HTTP request failed
+	 */
+	public static Map<String, Object> request(OP op, String uri, Map<String, String> headers, Map<String, Object> body, ContentType contentType) throws Exception {
```
Do a type check for Object in the body

Eriten replied:
Let's write some documentation about what are the types of Map is accepted for body.

jmittert commented:
- Added a method to handle both POST and PUT requests
- Added documenation about the Map<String,Object> type
- Verified the JSON Map before using it
- Added a unit test for the verification function
- Whitespace fixes

Eriten commented on src/test/java/com/sugarcrm/candybean/webservices/WSSystemTest.java:
```diff
@@ -0,0 +1,293 @@
+		} catch (Exception e) {
+			Assert.fail(e.toString());
+		}
+		System.out.println(response.get("json"));
```
I think that this line is no longer needed.

Eriten commented:
If you have time, just remove the `System.out.println`. Other than that, I think the PR looks good to me.

jmittert commented:
Removed.
Running tests here: http://ci/job/Voodoo_CB_Custom_Tests/97/ and here http://ci/job/Voodoo_CB_Custom_Tests/98/
